torch/nn/parallel/distributed.py

```diff
@@ -563,6 +563,12 @@ def forward(self, *inputs, **kwargs):
         if self.require_forward_param_sync:
             self._sync_params()
 
+        # It may allocate new buckets before deallocating old buckets
+        # inside rebuildBuckets.
+        # To save peak memory usage, call rebuildBuckets before the peak
+        # memory usage increases during computing outputs.
+        self.reducer.rebuildBuckets()
```

Comment:
Is it true that call this in the first fired-fired autograd hook is already too late?

Reply:
autograd hook will still record indices of grads in dynamic order, rebuildBuckets here is to use those recorded indices to build new buckets, so it should be fine to call it here